### ISSUE TITLE
generalize to arbitrary buffer sizes

### ISSUE THREAD

**Issue by afilinski** [Ginko-X/Streaming_NESL#18]
* once runnable proc selected, should step it for long as possible (until input buffer runs empty, or outbuffer full)
* buffer "stealing" (prematurely switch from filling to draining mode)
    * may sometimes be needed, but shouldn't be done otherwise (i.e., when some other process is runnable)

**Comment by Ginko-X** [comment]
- done, see the working progress of this week
- have several questions: 
    1.    
    > may sometimes be needed, 
    
    when? As I remember, we talked about it as an approach to unlock a deadlock. But I can't figure out a situation where we can use it to unlock a deadlock. As for now, all the deadlock cases I've seen are caused by attempting to traverse one stream >1 times at the same time, such as `let x= &15 in x++x`, `let x = &15 in concat({x}) ++ x` (assuming buffer size 10). And stealing should not be helpful to deal with this kind of deadlock 
    2. in my implementation for bullet one, Procs switches from filling to draining mode automatically in some cases. One is when the size of a Proc's all outputs is less than the buffer size, e.g., when evaluating `&4` with a  buffer size 10, no Procs can reach a full-filling state, but the interpreting can still move on to give us `{0,1,2,3}`. Another is when some Proc (always) outputs less than its input, such as `Usum`, `Pack`.  
    So maybe this is what you really meant?

**Comment by afilinski** [comment]
1. The probably simplest case where buffer stealing is needed is a situation like

```
let s = &100;
    t = {f(x) : x in s}
    u = {g(x) : x in s}
    r = {x+y : x in t; y in u}
```
when the buffers of t and u are not the same size. (In Frederik's SNESL implementation, we allocate a fixed number of bytes per buffer, as opposed to a fixed number of elements, since that's what is important for keeping the buffers in cache. Thus, buffers for boolean streams hold significantly more elements than integer streams.) Suppose the buffer for t can hold 10 elements and the one for u can hold 20. Then once t fills up, it'll switch to draining mode, but u is still in filling mode, and that means that r is stuck waiting for u. On the other hand, s cannot produce any more elements (which would help fill u's buffer), because t cannot process any more until it's output has been read. Only switching u's buffer to drain mode, even though it's still only half full, will break the deadlock.

Even if all buffers are sized to contain the same number of elements, similar situations arise. For example, consider
```
let s = &100
    r = {if x%10 == 0 then x/3 else x*3 : x in s}
```
Even if all buffers are the same size, the one for the multiplication process is going to fill up much quicker than the one for division, because it has to process 10 times as many elements. So again, to be able to produce r in constant buffer space, we'll have to regularly invoke the division process with only a barefly-full input buffer. [Incidentally, you should add the '%' operator to your implementation.]

2. Looking at your code, I don't see where/how you implement the switching between filling and draining  at all. I would have expected something like **(note fix in line 2)**:

```
data BufState =
    Filling [AVal] -- list contains less than (not "at most") bufSize elements
  | Draining [AVal]  -- list is non-empty, and with at most bufSize elements
  | Eos

```
so that you can distinguish between a half-full buffer in filling mode vs. draining mode. (Though it is possible that you somehow maintain that same information using just the read cursors, in which case it should be more clearly documented.)

It's almost certainly not a good idea to automatically switch a buffer to draining mode before it's full (unless, of course, you have definitely reached the end of the stream), because that will introduce many more context switches than necessary. For example, the SNESL code

`{reducePlus(&x) : x in &n)`

should be executable in a constant number of steps (just like in the SNESL interpreter), assuming the buffer size is sufficiently large (say, >= n²). Switching the output buffer of the &x to draining mode after processing each x will require a number of steps linear in n, or worse.

**Comment by Ginko-X** [comment]
1. I may understand what you mean. Can we now just ignore the hardware effect, and only consider all buffers are sized to contain the same number of elements? Then I think a concrete example may be like:
```Haskell
-- buffer size 1
let s = {1,3,5,7,2,4,6,8};
    t = concat({{x| x % 2 == 0 }: x in s});  -- even numbers
    u = concat({{x| x % 2 > 0 }: x in s})   -- odd ones
in {a + b : a in t, b in u}
```
I agree that this deadlock should be able to be broken by stealing.

The other example is confusing:
```
let s = &100
    r = {if x%10 == 0 then x/3 else x*3 : x in s}
```
I think you might mean `let s = &100 in {if x%10 == 0 then x/3 else x*3 : x in s}`,  then it seems not supposed to deadlock in any case, as it's just like a mapping operation on `s` by traversing `s` only once.

2.  I might haven't explained my implementation correctly:
> 2. in my implementation for bullet one, Procs switches from filling to draining mode automatically in some cases. 

What I did in my code is just extending the buffer from size 1 to `bufSize`. The SVCODE Procs are still scheduled in a round robin fashion, and in each round the Proc will keep transducing until it has read all the input elements (read blocking) or its buffer is full (write blocking). But in every new round, the buffer can be completely updated (no matter whether it is full or not) as long as all its clients have read all of the elements in previous rounds.

As you reminded me, I didn't consider if this scheduling algorithm has any problems with the cost model/space usage. I'll review the code and make necessary modifications.

**Comment by afilinski** [comment]
(1.) Stealing makes no difference for buffer size = 1, because in that case, the buffer gets switched to draining mode anyway, as soon as an element is added to it. Your example does not demonstrate an opportunity for stealing, but rather that a SNESL program can deadlock with a small buffer, even though it would have succeeded with a larger one (or an unbounded one, like in the eager interpreter), and that's a separate issue.

The reason why I brought up booleans vs. ints was just that in general, there's no need to require that all buffers have the same size (in number of elements), since making that assumption doesn't really simplify the interpreter: stealing may still be needed, as in my second example. Frederik and I even thought about letting the individual buffer sizes auto-adjust during execution (whenever they had been completely drained) in an adaptive way, to tune the computation for best performance while minimizing the space usage. We never implemented that, though.

And as for the second example, yes, to make it a complete SNESL program, either the second comprehension should be the body of the let-expression, or one should add "... in r". And my point was precisely that it _should not_ deadlock, but that it actually _will_, if the streaming interpreter naively insists on always filling up buffers completely (or reaching an EOS) before they can be switched to draining mode. But it should still try to fill up buffers whenever possible, to keep the number of steps low, i.e., to do as much work as possible in each step. In the example, the Mul transducer (where most of the work happens) will always run with a full buffer, so it doesn't matter asymptotically that the Mul steps alternate with some almost-empty Div steps.

(2.) I think it might be a good idea to add back the work/step counters (#19) now, so that you can easily see if the number of steps grows unexpectedly high in some cases. Also, to ease experimentation, you should probably make the buffer size interactively configurable, with a "`:bs <int>`" command, or similar. You already have some persistent state in InterEnv, so it should just be a matter of adding another component to that.

**Comment by Ginko-X** [comment]
1. This is tricky. So the buffer size you are talking about actually refers to **the number of bytes** in the real implementation of SNESL, not **the number of elements** at the high level, and the deadlock case for stealing to handle is caused by the unbalanced space usage of the buffers?

What I thought about with my example is that when the execution deadlocks as the following case shows,
```Haskell
-- buffer size 2; any larger size will not deadlock
let s = {1,3, 5, 2, 4, 6}; 
    t = concat({{x| x%2==0 }: x in s});  -- even numbers
    u = concat({{x| x%2 > 0 }: x in s})   -- odd numbers
in {a + b : a in t, b in u}
```
```Haskell
-- execution context  when it deadlocks
...
S15 := (Buf [5,2], [...],  [...],   <Pout 4 <Pin 9 <Proc>>> )  -- s: {1,3,5,2,4,6} 
...
S23 := (Buf [2],   [..,S15], [((S36,0),1)],  <Pin 0 <Proc>>)   -- t : even numbers
...
S33 := (Buf [1,3], [..,S15], [((S36,1),1)],  <Pout 5 <Pin 0 <Proc>>> )  -- u: odd numbers 
... 
S36 := (Buf [3],   [S23,S33], [((S-1,0),1)],  <Pin 0 <Proc>>)  -- final output: MapTwo Plus S23 S33
```

and now if we allow `S23` (as well as `S15`, `S33` and other related streams) to drain, then `S36` may be able to continue to produce, and then deadlock is broken. So seems I went a more complicated and wrong way. 

2. I'll try to add back cost counters first for the next step of work, then see what kind of problems I need to fix.

**Comment by afilinski** [comment]
In your revised example, you already did some stealing; otherwise S36 could not even have produced the first 3, because to compute that, it must have read the 2 from S23's buffer, which is not yet full, and so should still be in filling mode. I think the problem may rather be that S33 is currently stuck on Pout 5, and it has not yet read the 2 from S15, so S15 can't be safely refilled with [4,6]. On the other hand, S36 has read the 2 from S23 and the 1 from S33, and output 2+1=3, so now it's waiting for the next element from S23, which should be a 4. But S23 can only produce a 4 after reading it from S15, and S15 is still waiting for S33 to read the 2. So this is a real deadlock, as it should be: if the input in is bunched up with all the odd numbers first, it's impossible to compute the result in constant space, and you need a buffer of size at least half the length of s.

**Addendum:**  counting buffer sizes in bytes vs. elements is actually not important in itself; the point is that imposing the requirement that all buffers must contain the same number of elements does not actually avoid the need for stealing (i.e., switching a half-full buffer from filling to draining) in general, so  for the purpose of analyzing scheduling strategies, one might as well assume that all buffers can have different sizes, possibly determined by hardware considerations or even runtime profiling.

**Comment by Ginko-X** [comment]
1. One more example I'd like to discuss with you tomorrow (from Frederik's PhD thesis Page 112):

![image](https://user-images.githubusercontent.com/5687628/28279516-ffda835a-6b20-11e7-942a-57f2eb75199d.png)
![image](https://user-images.githubusercontent.com/5687628/28279682-86f14a54-6b21-11e7-9acb-e928979eba94.png)

He also mentioned:
> As an example, in Figure 4.3, notice that there are two strategies for further execution: Either fire the consumer(s) of (C) even though its buffer is less than full, or repeat the steps in the figure in an attempt to fill the buffer completely. 

I think the latter strategy is what you expected to be implemented and the case for stealing to apply to if a deadlock occurs, while what I took in my code is the former one. 

2. Another question:
```
data BufState =
    Filling [AVal] -- list contains less than (not "at most") bufSize elements
  | Draining [AVal]  -- list is non-empty, and with at most bufSize elements
  | Eos
``` 
It seems when the buffer is filling to be full, its state should change to `Drainging `, as `Filling` contains < `bufSize` elements, not ≤ `bufSize` ? Similarly, when draining to 0 element, the state becomes `Filling`, as `Draining` state must have > 0 element(s)?

3. I also remember you talked about something like "well-typed expressions never deadlock." So, if I recall correctly, what do you want to emphasize by saying so, or what do we need to do to achieve this?

**Comment by Ginko-X** [comment]
A fixed implementation has been updated. 
Here are some key notes:

1. The new `BufState` uses the same structure as you suggested, but no clients in draining state, i.e., not `Draining [AVal] Clients`. The reason is when switches mode (from filling to draining), the information of clients still needs to be obtained from context, so may not bring so much convenience compared to maintaining clients in each stream state.
```Haskell
data BufState = Filling [AVal] 
              | Draining [AVal]
              | Eos
```

2.  Interpreting a stream instruction in each scheduling round:
 (two-phase optimization hasn't been added)
The code looks even more complicated as now there are in total 3(bufState)*3(Proc types)  = 9 cases to analyse.

```Haskell
-- return a boolean to indicate if the stream proc has done
sInstrInterp :: Int -> SInstr -> SvcodeP Bool
sInstrInterp bufSize def@(SDef sid i) = 
  do (buf, sups, bs, p) <- lookupSid sid 
     case p of 
       -- 1. Done 
       Done () -> 
         case buf of 
           -- 1.1 the last chunk, so must use stealing 
           Filling as -> let buf' = if null as then Eos else Draining as 
                         in do updateCtx sid (buf', sups, bs, Done()) 
                               return True
           
           -- 1.2 just wait all clients finish reading its buffer
           Draining as -> if allEnd bs (length as)
                          then do updateCtx sid (Eos, sups, resetCur bs, Done ()) 
                                  returnC (length as,1) True  -- add cost 
                          else return True

           -- 1.3 do nothing
           Eos -> return True

       -- 2. Pout
       Pout a p' ->  
         case buf of 
           -- 2.1 fill the buffer until full 
           -- and then switch to draining 
           Filling as -> do let buf' = if length as +1 >= bufSize  
                                         then Draining $ as ++ [a] 
                                         else Filling $ as ++ [a]
                            updateCtx sid (buf', sups, bs, p') 
                            sInstrInterp bufSize def  

           -- 2.2 only check if all clients have finished reading the buffer
           -- if so, switch to filling
           Draining as -> 
             if allEnd bs (length as)
               then do updateCtx sid (Filling [], sups, resetCur bs, Pout a p')
                       costInc (bufSize, 1)  -- cost work `bufSize`, step 1
                       sInstrInterp bufSize def -- start filling
               else return False -- blocking 

           -- 2.3 can not happen
           Eos -> fail $ "Premature EOS "
       
-- 3. Pin i p'
-- for simplicity, only try to read from the supplier and modify the cursor if
-- read successfully (so not necessary to distinguish between 3.2 & 3.3)
       Pin i p' -> 
         case buf of 
           Eos -> fail $ "Premature EOS"  -- 3.1

           _ ->   -- 3.2 & 3.3
             do (bufSup, supSup, flagSup,pSup) <- lookupSid (sups!!i)
                case lookup (sid,i) flagSup of -- find own cursor from supplier's clients
                  Nothing -> fail $ "undefined client " ++ show sid
                  Just cursor -> 
                    case bufSup of -- check the bufState of the supplier
                      -- read a `Nothing`
                      Eos -> do updateCtx sid (buf,sups,bs, p' Nothing) 
                                sInstrInterp bufSize def  -- loop 

                      -- read unavailable, blocking
                      Filling _ -> return False  
                      
                      Draining asSup -> 
                        if cursor >= length asSup
                          then return False -- read blocking 
                          else -- read successfully
                            do let flagSup' = markRead flagSup (sid,i) cursor
                               updateCtx (sups!!i) (bufSup, supSup, flagSup',pSup)  
                               updateCtx sid (buf,sups,bs,p' (Just $ asSup !! cursor))
                               costInc (1,0)  -- cost work 1, step 0
                               sInstrInterp bufSize def -- loop
```

3. Stealing
When deadlock is detected (by comparing the contexts before and after a scheduling round), we'll first try to use stealing to break it. The implementation idea is to find the first non-empty stream in Filling mode and switch it to Draining mode. 
```Haskell
stealing :: Svctx -> Either String Svctx
stealing [] = Left "Deadlock!"
stealing ((sid, (Filling as@(a0:_), sup, bs, p)):ss) = 
  Right $ (sid, (Draining as, sup,bs,p)) : ss 
stealing (s:ss) = stealing ss >>= (\ss' -> return (s:ss'))
```
In many cases, just switching one stream is not enough. If no Filling stream exists, and the deadlock still can not be broken, then it should be a true deadlock. (??)

With regard to cost model, stealing should not affect it too much, (but I haven't  found enough argument to support it.)

4. WithCtrl seems work well, only small change from previous code for the bufSize-1 interpreter.

**Comment by afilinski** [comment]
random notes from meeting today:

transducer (xducer) (formerly "proc"):
  - streamble specification of computation, always one elt at a time
  - xducer can be (1) requesting input, (2) requesting output, or (3) done
  - usable by both eager and streaming interpreter

Process:
  consists of xducer (w/ internal state) and output buffer with associated state
  Once proc scheduled for execution (woken/activated), should step transducer
    as long as possible until it blocks (= requesting I/O that cannot be
      fulfilled without running some other process) or done.
       - when blocking, should identify cause (as hint for scheduler)
   - eos should be handled in same activation, so only one activation per
      process needed for sufficiently large buffers
  process may be (1) runnable, (2) blocked on input (from specific supplier),
    (3) blocked on output (because output buffer draining), or (4) finished
    - not explicitly a part of process state, but expressed by which list
      the process is on. (finished processes are not on any list)
    - should be redudant: waking a process with nothing to do should
        be harmless; it'll just go back to sleep immediately 
      - so trivial scheduler can just activate all non-finished processes in
         turn, but that makes it hard to detect deadlock and assign
	 step costs.
      
output buffer may be
    - filling: non-full list of elts (or array + write cursor), list of
       blocked clients
    - draining: non-empty list of elts, eos flag (true flag counts as non-empty),
      blocked producer process?
    - eos: no longer active, all clients have received eos (so will not ask again)
      - no longer any associated process. (can be GC'ed?)
      - each time last client finishes draining the buffer, use
        eos flag to determine whether new state is Filling [] or Eos.

  when switching from draining to filling, unblock producer proc,
    - happens when last client reads last elt, and not eos yet
  when switching from filling to draining, unblock all client procs
   - can happen either if buffer full, transducer done, or stealing
   
initially process 0 (ctrl) is only process on ready list, all others waiting
  for input on first port to read from (should always be port 0?)

scheduler runs through ready list one by one, constructs ready list for
next step, and puts previously ready processes on appropriate waiting lists
  - simd mode: each elt on ready list counts 1 step
  - mimd mode: 1 step reagardless of length of ready list
  - sanity check: previous and new ready lists are disjoint. (or not?)

when new ready list is empty, could be either that everything finished,
  or deadlock; need to distinguish somehow (counting #active procs?)
deadlock may be:
  - hard (fatal)
  - soft (resolvable, at least temporarily, by stealing)
